compiler/daml-lf-verify/src/DA/Daml/LF/Verify/Context.hs

```diff
@@ -0,0 +1,138 @@
+import Data.HashMap.Strict (HashMap, union, empty)
+
+import DA.Daml.LF.Ast
+import DA.Daml.LF.TypeChecker.Error
+
+data UpdCreate = UpdCreate
+  { _creTemp  :: !(Qualified TypeConName)
+    -- ^ Qualified type constructor corresponding to the contract template.
+  , _creField :: ![(FieldName, Expr)]
+    -- ^ The fields to be verified, together with their value.
+  }
+data UpdArchive = UpdArchive
+  { _arcTemp  :: !(Qualified TypeConName)
+    -- ^ Qualified type constructor corresponding to the contract template.
+  , _arcField :: ![(FieldName, Expr)]
+    -- ^ The fields to be verified, together with their value.
+  }
+data UpdChoice = UpdChoice
+  { _choTemp  :: !(Qualified TypeConName)
+    -- ^ Qualified type constructor corresponding to the contract template.
+  , _choName  :: !ChoiceName
+    -- ^ The name of the choice.
+  }
+
+-- | The List of updates being performed
+data UpdateSet = UpdateSet
+  { _usCre :: ![UpdCreate]
+    -- ^ The list of create updates.
+  , _usArc :: ![UpdArchive]
+    -- ^ The list of archive updates.
+  , _usCho :: ![UpdChoice]
+    -- ^ The list of choice updates.
+  }
+
+makeLenses ''UpdateSet
+
+emptyUpdateSet :: UpdateSet
+emptyUpdateSet = UpdateSet [] [] []
+
+concatUpdateSet :: UpdateSet -> UpdateSet -> UpdateSet
+concatUpdateSet (UpdateSet cres1 arcs1 chos1) (UpdateSet cres2 arcs2 chos2) =
+  UpdateSet (cres1 ++ cres2) (arcs1 ++ arcs2) (chos1 ++ chos2)
+
+-- | The environment for the DAML-LF verifier
+data Delta = Delta
+  { _devars :: ![ExprVarName]
+    -- ^ The skolemised term variables.
+  , _devals :: !(HashMap (Qualified ExprValName) (Expr, UpdateSet))
+    -- ^ The bound values.
+  , _dchs :: !(HashMap ((Qualified TypeConName), ChoiceName) UpdateSet)
+    -- ^ The set of relevant choices.
+  -- TODO: split this off into data types for readability?
+  }
+
+makeLenses ''Delta
+
+emptyDelta :: Delta
+emptyDelta = Delta [] empty empty
+
+concatDelta :: Delta -> Delta -> Delta
+concatDelta (Delta vars1 vals1 chs1) (Delta vars2 vals2 chs2) =
+  Delta (vars1 ++ vars2) (union vals1 vals2) (union chs1 chs2)
+  -- TODO: union makes me slightly nervous, as it allows overlapping keys
+  -- (and just uses the first)
+
+-- | Type class constraint with the required monadic effects for functions
+-- manipulating the verification environment.
+type MonadDelta m = (MonadError Error m, MonadReader Delta m)
+
+runDelta :: ReaderT Delta (Either Error) a -> Either Error a
+runDelta act = runReaderT act emptyDelta
+
+-- | Run a computation in the current environment, with an additional
+-- environment extension.
+introDelta :: MonadDelta m => Delta -> m a -> m a
+introDelta delta = local (concatDelta delta)
+
+extVarDelta :: MonadDelta m => ExprVarName -> m a -> m a
+extVarDelta x = local (over devars ((:) x))
+
+lookupDExprVar :: MonadDelta m => ExprVarName -> m ()
+lookupDExprVar x = ask >>= \ del -> unless (elem x $ del ^. devars)
+                                          $ throwError $ EUnknownExprVar x
```

Comment:
I don't think lenses buy us anything here. I also think `a >>= \x -> ...` is a sign to use `do` notation.
```suggestion
lookupDExprVar x = do
  del <- ask
  unless (x `elem` _devars del) $ throwError $ EUnknownExprVar x
```